ogr/services/pagure.py

```diff
@@ -127,7 +127,12 @@ def __hash__(self) -> int:
         return hash(str(self))
 
     def get_project(self, **kwargs) -> "PagureProject":
-        return PagureProject(service=self, **kwargs)
+        if "username" in kwargs:
+            return PagureProject(service=self, **kwargs)
+        else:
+            return PagureProject(
+                service=self, username=self.user.get_username(), **kwargs
+            )
```

Comment:
What do you think about this solution?

(Just a suggestion. Pick the one you like more.)

```python
    def get_project(self, username=None, **kwargs) -> "PagureProject":
        username = username or self.user.get_username()
        return PagureProject(
            service=self, username=self.user.get_username(), **kwargs
        )
```
----------

Btw, these two code snippets are equivalent..;-)

```python
if something:
    return "a"
else:
    return "b"
```
```python
if something:
    return "a"
return "b"
```

Reply:
Yeah, I know...
Why does it remind me of JS...

Reply:
Yes..;)  However, those oneliners usually bring no readability benefit.

Reply:
also that rewrite broke the tests...

Reply:
I do not see the difference.

Reply:
thought it's side effect of `or` evaluating expression before checking first one... tested in REPL, should not be

Reply:
```python
In [1]: def fce(): 
   ...:     print("hello") 
   ...:     return "a" 
   ...:                                                                                                                                                                                                                                        

In [2]: a = "a" or fce()                                                                                                                                                                                                                       

In [3]: a = None or fce()                                                                                                                                                                                                                      
hello
```

Sorry, but I don't understand.

Only difference I can see is when in `kwargs` you have `'username': None`.

Reply:
your suggestion breaks 4 tests

```python
username = kwargs["username"] if "username" in kwargs else self.user.get_username()
return PagureProject(
  service=self, username=username, **kwargs
)
```
this breaks 10...

Reply:
Also what I've noticed, there are some intentionally failing tests for `get_project` like getting github project out of it... I don't understand why they were fine with `if ... else ...` but not with that

Reply:
My fault, I had an error in my code -- I've forgotten to use the username variable in the PagureProject call. (I've fixed that in the comment.)

Reply:
I was using the username

Reply:
Hmm, that's weird.

> atter one has the problem with multiple values for 'username'

How is that possible?

I'll try to play with your code locally to see what is happening there.

Reply:
In the meantime, go with the solution that works better for you.

Reply:
> How is that possible?

this one (the one that broke the record of broken tests):
```python
username = kwargs["username"] if "username" in kwargs else self.user.get_username()
return PagureProject(
  service=self, username=username, **kwargs
)
```
it does pass it two times, once via username and second time with `**`

Reply:
this works:
```python
def get_project(self, username=None, **kwargs) -> "PagureProject":
    if username:
        return PagureProject(
            service=self, username=username, **kwargs
        )

    return PagureProject(
        service=self, **kwargs
    )
```
and this doesn't:
```python
def get_project(self, username=None, **kwargs) -> "PagureProject":
    if username:
        return PagureProject(
            service=self, username=username, **kwargs
        )

    return PagureProject(
        service=self, username=self.user.get_username(), **kwargs
    )
```

Also it appears to cause problems with "invalid" tests:
1. trying to get project from URL with `github.com` repository
1. `pagure.io/ogr` -> non-existing
1. `src.fedoraproject.org` didn't make token for that, so that could be reason of failing on my side

Reply:
> Also it appears to cause problems with "invalid" tests:

It started to touch the remote services what was not expected when writing the code. (The test uses mapping like `{"github.com": PagureService}` what is fine until we try to call an API.)

Let's use the current approach and spend some time on `offline x online` handling later.